"""
–¢–µ—Å—Ç–æ–≤—ã–π Python gRPC —Å–µ—Ä–≤–∏—Å –¥–ª—è –≥–µ–æ–∫–æ–¥–∏—Ä–æ–≤–∞–Ω–∏—è
–û–±–Ω–æ–≤–ª–µ–Ω–æ –ø–æ–¥ –Ω–æ–≤—É—é —Å—Ç—Ä—É–∫—Ç—É—Ä—É API
"""
import logging
import time
from concurrent import futures
import sys
import io
import grpc
import geocode_pb2
import geocode_pb2_grpc

# –ò—Å–ø—Ä–∞–≤–ª—è–µ–º –∫–æ–¥–∏—Ä–æ–≤–∫—É –¥–ª—è Windows
if sys.platform == 'win32':
    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')
    sys.stderr = io.TextIOWrapper(sys.stderr.buffer, encoding='utf-8')

# –ù–∞—Å—Ç—Ä–æ–π–∫–∞ –ª–æ–≥–∏—Ä–æ–≤–∞–Ω–∏—è
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class GeocodeServicer(geocode_pb2_grpc.GeocodeServiceServicer):
    """
    –†–µ–∞–ª–∏–∑–∞—Ü–∏—è gRPC —Å–µ—Ä–≤–∏—Å–∞ –≥–µ–æ–∫–æ–¥–∏—Ä–æ–≤–∞–Ω–∏—è
    """

    def __init__(self):
        self.start_time = time.time()
        # –¢–µ—Å—Ç–æ–≤—ã–µ –¥–∞–Ω–Ω—ã–µ (–≤ —Ä–µ–∞–ª—å–Ω–æ–π –≤–µ—Ä—Å–∏–∏ –±—É–¥–µ—Ç –±–∞–∑–∞ –¥–∞–Ω–Ω—ã—Ö)
        self.mock_data = [
            {
                "locality": "–ú–æ—Å–∫–≤–∞",
                "street": "–¢–≤–µ—Ä—Å–∫–∞—è —É–ª–∏—Ü–∞",
                "number": "7",
                "lon": 37.615560,
                "lat": 55.757814,
                "score": 0.95,
                "postal_code": "125009",
                "district": "–¢–≤–µ—Ä—Å–∫–æ–π —Ä–∞–π–æ–Ω",
                "full_address": "–ú–æ—Å–∫–≤–∞, –¢–≤–µ—Ä—Å–∫–∞—è —É–ª–∏—Ü–∞, 7"
            },
            {
                "locality": "–ú–æ—Å–∫–≤–∞",
                "street": "–ö—Ä–∞—Å–Ω–∞—è –ø–ª–æ—â–∞–¥—å",
                "number": "1",
                "lon": 37.621211,
                "lat": 55.753544,
                "score": 0.98,
                "postal_code": "109012",
                "district": "–¢–≤–µ—Ä—Å–∫–æ–π —Ä–∞–π–æ–Ω",
                "full_address": "–ú–æ—Å–∫–≤–∞, –ö—Ä–∞—Å–Ω–∞—è –ø–ª–æ—â–∞–¥—å, 1"
            },
            {
                "locality": "–ú–æ—Å–∫–≤–∞",
                "street": "–ø—Ä–æ—Å–ø–µ–∫—Ç –ú–∏—Ä–∞",
                "number": "119",
                "lon": 37.639600,
                "lat": 55.822144,
                "score": 0.92,
                "postal_code": "129223",
                "district": "–û—Å—Ç–∞–Ω–∫–∏–Ω—Å–∫–∏–π —Ä–∞–π–æ–Ω",
                "full_address": "–ú–æ—Å–∫–≤–∞, –ø—Ä–æ—Å–ø–µ–∫—Ç –ú–∏—Ä–∞, 119"
            },
            {
                "locality": "–°–∞–Ω–∫—Ç-–ü–µ—Ç–µ—Ä–±—É—Ä–≥",
                "street": "–ù–µ–≤—Å–∫–∏–π –ø—Ä–æ—Å–ø–µ–∫—Ç",
                "number": "28",
                "lon": 30.324116,
                "lat": 59.935493,
                "score": 0.90,
                "postal_code": "191186",
                "district": "–¶–µ–Ω—Ç—Ä–∞–ª—å–Ω—ã–π —Ä–∞–π–æ–Ω",
                "full_address": "–°–∞–Ω–∫—Ç-–ü–µ—Ç–µ—Ä–±—É—Ä–≥, –ù–µ–≤—Å–∫–∏–π –ø—Ä–æ—Å–ø–µ–∫—Ç, 28"
            },
            {
                "locality": "–ú–æ—Å–∫–≤–∞",
                "street": "—É–ª–∏—Ü–∞ –ê—Ä–±–∞—Ç",
                "number": "10",
                "lon": 37.593434,
                "lat": 55.750446,
                "score": 0.88,
                "postal_code": "119019",
                "district": "–ê—Ä–±–∞—Ç",
                "full_address": "–ú–æ—Å–∫–≤–∞, —É–ª–∏—Ü–∞ –ê—Ä–±–∞—Ç, 10"
            },
        ]

    def SearchAddress(self, request, context):
        """
        –û–±—Ä–∞–±–æ—Ç–∫–∞ –∑–∞–ø—Ä–æ—Å–∞ –Ω–∞ –ø–æ–∏—Å–∫ –∞–¥—Ä–µ—Å–∞
        """
        start_time = time.time()

        # –î–µ—Ç–∞–ª—å–Ω–æ–µ –ª–æ–≥–∏—Ä–æ–≤–∞–Ω–∏–µ –≤—Ö–æ–¥—è—â–µ–≥–æ –∑–∞–ø—Ä–æ—Å–∞
        logger.info("=" * 80)
        logger.info(f"üì® –í–•–û–î–Ø–©–ò–ô –ó–ê–ü–†–û–° [request_id={request.request_id}]")
        logger.info("-" * 80)
        logger.info(f"üî§ –û—Ä–∏–≥–∏–Ω–∞–ª—å–Ω—ã–π –∑–∞–ø—Ä–æ—Å:     '{request.original_query}'")
        logger.info(f"‚ú® –ù–æ—Ä–º–∞–ª–∏–∑–æ–≤–∞–Ω–Ω—ã–π –∑–∞–ø—Ä–æ—Å:  '{request.normalized_query}'")
        logger.info(f"üìä –õ–∏–º–∏—Ç —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤:       {request.limit}")

        # –õ–æ–≥–∏—Ä–æ–≤–∞–Ω–∏–µ –æ–ø—Ü–∏–π –ø–æ–∏—Å–∫–∞
        if request.options:
            logger.info(f"‚öôÔ∏è  –û–ø—Ü–∏–∏ –ø–æ–∏—Å–∫–∞:")
            logger.info(f"   ‚Ä¢ min_score_threshold:  {request.options.min_score_threshold}")
            logger.info(f"   ‚Ä¢ enable_fuzzy_search:  {request.options.enable_fuzzy_search}")
            if request.options.locality_filter:
                logger.info(f"   ‚Ä¢ locality_filter:      '{request.options.locality_filter}'")

        # –õ–æ–≥–∏—Ä–æ–≤–∞–Ω–∏–µ —Å—Ç—Ä—É–∫—Ç—É—Ä–∏—Ä–æ–≤–∞–Ω–Ω—ã—Ö –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤ –∞–¥—Ä–µ—Å–∞
        if request.parsed_components:
            components = request.parsed_components
            logger.info(f"üèóÔ∏è  –°—Ç—Ä—É–∫—Ç—É—Ä–∏—Ä–æ–≤–∞–Ω–Ω—ã–µ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã (libpostal):")

            # –û—Å–Ω–æ–≤–Ω—ã–µ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã
            if components.city:
                logger.info(f"   ‚Ä¢ –ì–æ—Ä–æ–¥:                '{components.city}'")
            if components.road:
                logger.info(f"   ‚Ä¢ –£–ª–∏—Ü–∞:                '{components.road}'")
            if components.house_number:
                logger.info(f"   ‚Ä¢ –ù–æ–º–µ—Ä –¥–æ–º–∞:           '{components.house_number}'")

            # –î–æ–ø–æ–ª–Ω–∏—Ç–µ–ª—å–Ω—ã–µ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã
            if components.unit:
                logger.info(f"   ‚Ä¢ –ö–≤–∞—Ä—Ç–∏—Ä–∞/–û—Ñ–∏—Å:        '{components.unit}'")
            if components.level:
                logger.info(f"   ‚Ä¢ –≠—Ç–∞–∂:                 '{components.level}'")
            if components.staircase:
                logger.info(f"   ‚Ä¢ –ü–æ–¥—ä–µ–∑–¥:              '{components.staircase}'")
            if components.entrance:
                logger.info(f"   ‚Ä¢ –í—Ö–æ–¥:                 '{components.entrance}'")

            # –ê–¥–º–∏–Ω–∏—Å—Ç—Ä–∞—Ç–∏–≤–Ω—ã–µ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã
            if components.suburb:
                logger.info(f"   ‚Ä¢ –†–∞–π–æ–Ω/–ú–∏–∫—Ä–æ—Ä–∞–π–æ–Ω:     '{components.suburb}'")
            if components.city_district:
                logger.info(f"   ‚Ä¢ –û–∫—Ä—É–≥ –≥–æ—Ä–æ–¥–∞:         '{components.city_district}'")
            if components.state:
                logger.info(f"   ‚Ä¢ –†–µ–≥–∏–æ–Ω/–û–±–ª–∞—Å—Ç—å:       '{components.state}'")
            if components.country:
                logger.info(f"   ‚Ä¢ –°—Ç—Ä–∞–Ω–∞:               '{components.country}'")

            # –ü–æ—á—Ç–æ–≤—ã–π –∏–Ω–¥–µ–∫—Å
            if components.postcode:
                logger.info(f"   ‚Ä¢ –ü–æ—á—Ç–æ–≤—ã–π –∏–Ω–¥–µ–∫—Å:      '{components.postcode}'")
        else:
            logger.info(f"‚ö†Ô∏è  –°—Ç—Ä—É–∫—Ç—É—Ä–∏—Ä–æ–≤–∞–Ω–Ω—ã–µ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã –æ—Ç—Å—É—Ç—Å—Ç–≤—É—é—Ç (fallback –Ω–∞ —Å—Ç—Ä–æ–∫–æ–≤—ã–π –ø–æ–∏—Å–∫)")

        logger.info("=" * 80)

        try:
            # –õ–æ–≥–∏–∫–∞ –ø–æ–∏—Å–∫–∞ —Å –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ–º —Å—Ç—Ä—É–∫—Ç—É—Ä–∏—Ä–æ–≤–∞–Ω–Ω—ã—Ö –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤
            components = request.parsed_components
            query_lower = request.normalized_query.lower()
            original_lower = request.original_query.lower() if request.original_query else query_lower
            results = []

            # –ò–∑–≤–ª–µ–∫–∞–µ–º –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã –∞–¥—Ä–µ—Å–∞ –¥–ª—è –±–æ–ª–µ–µ —Ç–æ—á–Ω–æ–≥–æ –ø–æ–∏—Å–∫–∞
            search_city = components.city.lower() if components and components.city else ""
            search_road = components.road.lower() if components and components.road else ""
            search_house = components.house_number.lower() if components and components.house_number else ""

            logger.debug(
                f"–ü–æ–∏—Å–∫ –ø–æ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞–º: city='{search_city}', road='{search_road}', house='{search_house}'"
            )

            for item in self.mock_data:
                # –ü—Ä–∏–æ—Ä–∏—Ç–µ—Ç 1: –ü–æ–∏—Å–∫ –ø–æ —Å—Ç—Ä—É–∫—Ç—É—Ä–∏—Ä–æ–≤–∞–Ω–Ω—ã–º –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞–º (–µ—Å–ª–∏ –µ—Å—Ç—å)
                if components and (search_city or search_road or search_house):
                    # –¢–æ—á–Ω–æ–µ —Å–æ–≤–ø–∞–¥–µ–Ω–∏–µ –ø–æ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞–º –¥–∞–µ—Ç –≤—ã—Å–æ–∫–∏–π score
                    city_match = search_city and search_city in item["locality"].lower()
                    road_match = search_road and search_road in item["street"].lower()
                    house_match = search_house and search_house in item["number"].lower()

                    if city_match or road_match or house_match:
                        # –ü–æ–≤—ã—à–∞–µ–º score –ø—Ä–∏ —Å–æ–≤–ø–∞–¥–µ–Ω–∏–∏ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤
                        adjusted_score = item["score"]
                        if city_match and road_match and house_match:
                            adjusted_score = min(1.0, item["score"] + 0.1)  # –í—Å–µ 3 –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞
                        elif (city_match and road_match) or (city_match and house_match):
                            adjusted_score = min(1.0, item["score"] + 0.05)  # 2 –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞

                        # –°–æ–∑–¥–∞–µ–º –¥–æ–ø–æ–ª–Ω–∏—Ç–µ–ª—å–Ω—É—é –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏—é
                        additional_info = geocode_pb2.AdditionalInfo(
                            postal_code=item.get("postal_code", ""),
                            district=item.get("district", ""),
                            full_address=item["full_address"],
                            object_id=f"obj_{len(results) + 1}"
                        )

                        # –°–æ–∑–¥–∞–µ–º –∞–¥—Ä–µ—Å–Ω—ã–π –æ–±—ä–µ–∫—Ç
                        address_object = geocode_pb2.AddressObject(
                            locality=item["locality"],
                            street=item["street"],
                            number=item["number"],
                            lon=item["lon"],
                            lat=item["lat"],
                            score=adjusted_score,
                            additional_info=additional_info
                        )

                        results.append(address_object)
                        continue

                # –ü—Ä–∏–æ—Ä–∏—Ç–µ—Ç 2: Fallback –Ω–∞ –ø–æ–∏—Å–∫ –ø–æ —Å—Ç—Ä–æ–∫–∞–º (–µ—Å–ª–∏ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã –Ω–µ –¥–∞–ª–∏ —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤)
                if (query_lower in item["locality"].lower() or
                    query_lower in item["street"].lower() or
                    query_lower in item["full_address"].lower() or
                    original_lower in item["locality"].lower() or
                    original_lower in item["street"].lower() or
                    original_lower in item["full_address"].lower()):

                    # –°–æ–∑–¥–∞–µ–º –¥–æ–ø–æ–ª–Ω–∏—Ç–µ–ª—å–Ω—É—é –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏—é
                    additional_info = geocode_pb2.AdditionalInfo(
                        postal_code=item.get("postal_code", ""),
                        district=item.get("district", ""),
                        full_address=item["full_address"],
                        object_id=f"obj_{len(results) + 1}"
                    )

                    # –°–æ–∑–¥–∞–µ–º –∞–¥—Ä–µ—Å–Ω—ã–π –æ–±—ä–µ–∫—Ç
                    address_object = geocode_pb2.AddressObject(
                        locality=item["locality"],
                        street=item["street"],
                        number=item["number"],
                        lon=item["lon"],
                        lat=item["lat"],
                        score=item["score"],
                        additional_info=additional_info
                    )

                    results.append(address_object)

            # –°–æ—Ä—Ç–∏—Ä—É–µ–º –ø–æ —Ä–µ–ª–µ–≤–∞–Ω—Ç–Ω–æ—Å—Ç–∏
            results.sort(key=lambda x: x.score, reverse=True)

            # –ü—Ä–∏–º–µ–Ω—è–µ–º —Ñ–∏–ª—å—Ç—Ä –ø–æ –º–∏–Ω–∏–º–∞–ª—å–Ω–æ–º—É score (–µ—Å–ª–∏ —É–∫–∞–∑–∞–Ω)
            if request.options and request.options.min_score_threshold > 0:
                results = [r for r in results if r.score >= request.options.min_score_threshold]

            # –û–≥—Ä–∞–Ω–∏—á–∏–≤–∞–µ–º –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤
            if request.limit > 0:
                results = results[:request.limit]

            execution_time_ms = int((time.time() - start_time) * 1000)

            # –î–µ—Ç–∞–ª—å–Ω–æ–µ –ª–æ–≥–∏—Ä–æ–≤–∞–Ω–∏–µ —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤
            logger.info("-" * 80)
            logger.info(f"‚úÖ –†–ï–ó–£–õ–¨–¢–ê–¢–´ –ü–û–ò–°–ö–ê [request_id={request.request_id}]")
            logger.info(f"   ‚Ä¢ –ù–∞–π–¥–µ–Ω–æ —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤:  {len(results)}")
            logger.info(f"   ‚Ä¢ –í—Ä–µ–º—è –≤—ã–ø–æ–ª–Ω–µ–Ω–∏—è:     {execution_time_ms}ms")

            if results:
                logger.info(f"   ‚Ä¢ –¢–æ–ø —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤:")
                for idx, result in enumerate(results[:3], 1):  # –ü–æ–∫–∞–∑—ã–≤–∞–µ–º —Ç–æ–ø-3
                    logger.info(
                        f"     {idx}. {result.locality}, {result.street}, {result.number} "
                        f"(score: {result.score:.2f})"
                    )
            else:
                logger.warning(f"   ‚ö†Ô∏è –ù–∏—á–µ–≥–æ –Ω–µ –Ω–∞–π–¥–µ–Ω–æ!")

            logger.info("=" * 80)

            # –í–æ–∑–≤—Ä–∞—â–∞–µ–º –æ—Ç–≤–µ—Ç (searched_address —Å–æ–¥–µ—Ä–∂–∏—Ç –Ω–æ—Ä–º–∞–ª–∏–∑–æ–≤–∞–Ω–Ω—ã–π –∑–∞–ø—Ä–æ—Å)
            return geocode_pb2.SearchAddressResponse(
                status=geocode_pb2.ResponseStatus(
                    code=geocode_pb2.StatusCode.OK,
                    message="–ü–æ–∏—Å–∫ –≤—ã–ø–æ–ª–Ω–µ–Ω —É—Å–ø–µ—à–Ω–æ"
                ),
                searched_address=request.normalized_query,
                objects=results,
                total_found=len(results),
                metadata=geocode_pb2.ResponseMetadata(
                    execution_time_ms=execution_time_ms,
                    timestamp=int(time.time()),
                    engine_version="1.0.0-mock"
                )
            )

        except Exception as e:
            logger.error("=" * 80)
            logger.error(f"‚ùå –û–®–ò–ë–ö–ê –ü–†–ò –û–ë–†–ê–ë–û–¢–ö–ï –ó–ê–ü–†–û–°–ê [request_id={request.request_id}]")
            logger.error(f"   ‚Ä¢ –¢–∏–ø –æ—à–∏–±–∫–∏: {type(e).__name__}")
            logger.error(f"   ‚Ä¢ –°–æ–æ–±—â–µ–Ω–∏–µ:  {str(e)}")
            logger.error("=" * 80, exc_info=True)
            return geocode_pb2.SearchAddressResponse(
                status=geocode_pb2.ResponseStatus(
                    code=geocode_pb2.StatusCode.INTERNAL_ERROR,
                    message=f"–í–Ω—É—Ç—Ä–µ–Ω–Ω—è—è –æ—à–∏–±–∫–∞ —Å–µ—Ä–≤–µ—Ä–∞",
                    details=str(e)
                ),
                searched_address=request.normalized_query,
                objects=[],
                total_found=0
            )

    def HealthCheck(self, request, context):
        """
        Health check –¥–ª—è –º–æ–Ω–∏—Ç–æ—Ä–∏–Ω–≥–∞
        """
        uptime = int(time.time() - self.start_time)

        return geocode_pb2.HealthCheckResponse(
            status=geocode_pb2.HealthStatus.HEALTHY,
            version="1.0.0",
            uptime_seconds=uptime
        )


def serve():
    """
    –ó–∞–ø—É—Å–∫ gRPC —Å–µ—Ä–≤–µ—Ä–∞
    """
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    geocode_pb2_grpc.add_GeocodeServiceServicer_to_server(
        GeocodeServicer(), server
    )

    # –°–ª—É—à–∞–µ–º –Ω–∞ –ø–æ—Ä—Ç—É 50051
    server.add_insecure_port('[::]:50051')
    server.start()

    logger.info("‚úÖ Python gRPC —Å–µ—Ä–≤–µ—Ä –∑–∞–ø—É—â–µ–Ω –Ω–∞ –ø–æ—Ä—Ç—É 50051")
    logger.info("üì° Endpoints:")
    logger.info("   - SearchAddress (–ø–æ–∏—Å–∫ –∞–¥—Ä–µ—Å–æ–≤)")
    logger.info("   - HealthCheck (–ø—Ä–æ–≤–µ—Ä–∫–∞ –∑–¥–æ—Ä–æ–≤—å—è)")
    logger.info("–û–∂–∏–¥–∞–Ω–∏–µ –∑–∞–ø—Ä–æ—Å–æ–≤...")

    try:
        server.wait_for_termination()
    except KeyboardInterrupt:
        logger.info("–û—Å—Ç–∞–Ω–æ–≤–∫–∞ —Å–µ—Ä–≤–µ—Ä–∞...")
        server.stop(0)


if __name__ == '__main__':
    serve()
